dev_guide/cicd.adoc

```diff
@@ -0,0 +1,254 @@
+consumed by those language specific runtimes.
+
+By pulling in OpenShift Origin, and its combined foundation of Kubernetes and Docker, additional application artfacts now minimally include:
+- Images with their rich set of metadata and associated tooling
+- Environment Variables and a powerful array of API Objects that are either injected into Containers and Pods for application use, or dictate how the Orchestration Engine (OpenShift on Kubernetes) manages
+those Containers and Pods
+
+And the second bullet there should serve as a reminder:  the lines between development and operations have blurred, and we want to improve the interaction between the two .... we now have "DevOps".
+
+But while the inventory of concerns has increased, many of the concepts and semantics around managing those concerns are thankfully very similar:
+- versioning
+- portability
+- isolation
+- upgrade and update
+- hooks (into various flows)
+
+We'll next map these concerns to what is either available in or accessible from OpenShift to this problem domain.
+
+[[versioning]]
+== Versioning
+
+Versioning in this topic's context refers to designating each of the components of your application to the stage in the `*CI/CD*` pipeline their readiness dictates they should reside.
+
+For your source code (in the classic sense), the OpenShift `*Build*`, `*BuildRequest*`, and `*BuildConfig*` object are of course the starting point.  They expose the most typically used
+features of the `*SCM*` (most notably Git) to access the appropriate version of your source code (commit IDs, branches, context directories, etc.) and build the source code.
+
+But source code management (`*SCM*`) tooling is also leveraged througout the industry for managing many artifacts beyond what classically is defined as source code.
+OpenShift also promotes such a strategy.
+
+And where a `*SCM*` is not appropriate, most notably Images, we have the OpenShift Internal Image Registry and `*ImageStreams*`. 
+
+[[api-objects]]
+=== API Objects
+
+As is highlighted throughout OpenShift documentation, every API Object can be expressed via either JSON or YAML.  That JSON or YAML can be stored in and retrieved from a
+`*SCM*` system, with the its versioning related capabilities, including the creation of branches, assignment of and query on various labels or tags associated to versions.
+
+Also, the API Ojects are designed such that there are portions of the object which are intended to *specify* the Object's intent, and other
+portions which reflect the *status* or current state of the Object.  Inputs and outputs in other words.  The input portions, when expressed in JSON/YAML, in particular are items
+that naturally fit as artifacts stored in a `*SCM*`.
+
+[NOTE]
+Remember, the input or specification portions of the API Objects can be totally static, or dynamic in the sense that variable substitution via template processing is possible on actual instantiation.
```

Comment:
s/static,/static/
The note seems a little out of place for me. I don't necessarily know what a template is.